File: mobile/types/billing.ts
/**
 * Billing and Subscription types - matching backend schemas/billing.py exactly
 */

import { SubscriptionTier, SubscriptionStatus } from './api';

// ===========================================
// Plan Feature
// ===========================================
export interface PlanFeature {
  name: string;
  description: string;
  included: boolean;
  limit?: string | null; // e.g., "100/month"
}

// ===========================================
// Pricing Plan
// ===========================================
export interface PricingPlan {
  id: string;
  name: string;
  tier: SubscriptionTier;
  description: string;
  price_monthly: number; // Price in cents
  price_yearly: number; // Price in cents (yearly total)
  stripe_price_id_monthly?: string | null;
  stripe_price_id_yearly?: string | null;
  features: PlanFeature[];
  is_popular: boolean;
  properties_limit: number;
  searches_per_month: number;
  api_calls_per_month: number;
}

// ===========================================
// Pricing Plans Response
// ===========================================
export interface PricingPlansResponse {
  plans: PricingPlan[];
}

// ===========================================
// Subscription Response
// ===========================================
export interface SubscriptionResponse {
  id: string;
  tier: SubscriptionTier;
  status: SubscriptionStatus;
  stripe_customer_id?: string | null;
  stripe_subscription_id?: string | null;
  current_period_start?: string | null;
  current_period_end?: string | null;
  cancel_at_period_end: boolean;
  canceled_at?: string | null;
  trial_start?: string | null;
  trial_end?: string | null;

  // Limits
  properties_limit: number;
  searches_per_month: number;
  api_calls_per_month: number;

  // Usage
  searches_used: number;
  api_calls_used: number;
  usage_reset_date?: string | null;

  created_at: string;
  updated_at: string;
}

// ===========================================
// Usage Response
// ===========================================
export interface UsageResponse {
  tier: SubscriptionTier;

  // Properties
  properties_saved: number;
  properties_limit: number;
  properties_remaining: number;

  // Searches
  searches_used: number;
  searches_limit: number;
  searches_remaining: number;

  // API calls
  api_calls_used: number;
  api_calls_limit: number;
  api_calls_remaining: number;

  // Reset info
  usage_reset_date?: string | null;
  days_until_reset?: number | null;
}

// ===========================================
// Checkout Request
// ===========================================
export interface CreateCheckoutRequest {
  price_id: string;
  success_url?: string | null;
  cancel_url?: string | null;
}

// ===========================================
// Checkout Session Response
// ===========================================
export interface CheckoutSessionResponse {
  checkout_url: string;
  session_id: string;
}

// ===========================================
// Portal Session Response
// ===========================================
export interface PortalSessionResponse {
  portal_url: string;
}

// ===========================================
// Cancel Subscription Request
// ===========================================
export interface CancelSubscriptionRequest {
  cancel_immediately: boolean;
  reason?: string | null;
}

// ===========================================
// Cancel Subscription Response
// ===========================================
export interface CancelSubscriptionResponse {
  success: boolean;
  message: string;
  cancel_at_period_end: boolean;
  current_period_end?: string | null;
}

// ===========================================
// Payment History Item
// ===========================================
export interface PaymentHistoryItem {
  id: string;
  amount: number; // In cents
  currency: string;
  status: string;
  description?: string | null;
  invoice_pdf_url?: string | null;
  receipt_url?: string | null;
  created_at: string;
}

// ===========================================
// Payment History Response
// ===========================================
export interface PaymentHistoryResponse {
  payments: PaymentHistoryItem[];
  total_count: number;
  has_more: boolean;
}

// ===========================================
// Helper Functions
// ===========================================

/**
 * Format price from cents to dollars string
 */
export function formatPrice(cents: number): string {
  return `$${(cents / 100).toFixed(2)}`;
}

/**
 * Format monthly price from yearly cents
 */
export function formatMonthlyFromYearly(yearlyCents: number): string {
  return `$${(yearlyCents / 100 / 12).toFixed(2)}`;
}

/**
 * Calculate savings percentage for yearly vs monthly
 */
export function calculateYearlySavings(monthlyPrice: number, yearlyPrice: number): number {
  const yearlyIfMonthly = monthlyPrice * 12;
  return Math.round(((yearlyIfMonthly - yearlyPrice) / yearlyIfMonthly) * 100);
}

File: backend/app/schemas/billing.py
"""
Pydantic schemas for billing and subscription management.
"""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field


class SubscriptionTier(StrEnum):
    """Subscription tier levels."""

    FREE = "free"
    PRO = "pro"


class SubscriptionStatus(StrEnum):
    """Subscription status."""

    ACTIVE = "active"
    PAST_DUE = "past_due"
    CANCELED = "canceled"
    INCOMPLETE = "incomplete"
    INCOMPLETE_EXPIRED = "incomplete_expired"
    TRIALING = "trialing"
    UNPAID = "unpaid"
    PAUSED = "paused"


# ===========================================
# Pricing Plans
# ===========================================


class PlanFeature(BaseModel):
    """A feature included in a plan."""

    name: str
    description: str
    included: bool = True
    limit: str | None = None  # e.g., "100/month"


class PricingPlan(BaseModel):
    """Pricing plan details."""

    id: str
    name: str
    tier: SubscriptionTier
    description: str
    price_monthly: int  # Price in cents
    price_yearly: int  # Price in cents (yearly total)
    stripe_price_id_monthly: str | None = None
    stripe_price_id_yearly: str | None = None
    features: list[PlanFeature]
    is_popular: bool = False
    properties_limit: int
    searches_per_month: int
    api_calls_per_month: int


class PricingPlansResponse(BaseModel):
    """Response with all pricing plans."""

    plans: list[PricingPlan]


# ===========================================
# Subscription Management
# ===========================================


class PlanType(StrEnum):
    """Billing interval for Pro (for display only)."""

    STARTER = "starter"
    PRO_MONTHLY = "pro_monthly"
    PRO_ANNUAL = "pro_annual"


class SubscriptionResponse(BaseModel):
    """Current subscription details."""

    id: str
    tier: SubscriptionTier
    status: SubscriptionStatus
    plan_type: PlanType | None = None  # starter | pro_monthly | pro_annual
    current_period_start: datetime | None = None
    current_period_end: datetime | None = None
    cancel_at_period_end: bool = False
    canceled_at: datetime | None = None
    trial_start: datetime | None = None
    trial_end: datetime | None = None

    # Limits
    properties_limit: int
    searches_per_month: int
    api_calls_per_month: int

    # Usage
    searches_used: int
    api_calls_used: int
    usage_reset_date: datetime | None = None

    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class UsageResponse(BaseModel):
    """Current usage statistics."""

    tier: SubscriptionTier

    # Properties
    properties_saved: int
    properties_limit: int
    properties_remaining: int

    # Searches
    searches_used: int
    searches_limit: int
    searches_remaining: int

    # API calls
    api_calls_used: int
    api_calls_limit: int
    api_calls_remaining: int

    # Reset info
    usage_reset_date: datetime | None = None
    days_until_reset: int | None = None


# ===========================================
# Checkout & Portal
# ===========================================


class CreateCheckoutRequest(BaseModel):
    """Request to create a Stripe checkout session."""

    price_id: str | None = Field(None, description="Stripe price ID (provide this or lookup_key)")
    lookup_key: str | None = Field(None, description="Stripe price lookup key (alternative to price_id)")
    success_url: str | None = Field(None, description="URL to redirect after success")
    cancel_url: str | None = Field(None, description="URL to redirect after cancel")


class CheckoutSessionResponse(BaseModel):
    """Stripe checkout session response."""

    checkout_url: str
    session_id: str


class SetupIntentResponse(BaseModel):
    """Stripe SetupIntent client secret for embedded card collection."""

    client_secret: str


class CreateSubscriptionRequest(BaseModel):
    """Request to create a subscription after collecting payment method."""

    payment_method_id: str = Field(..., description="Stripe PaymentMethod ID from confirmed SetupIntent")
    price_id: str | None = Field(None, description="Stripe price ID")
    lookup_key: str | None = Field(None, description="Stripe price lookup key")


class CreateSubscriptionResponse(BaseModel):
    """Response after subscription creation."""

    subscription_id: str
    status: str
    trial_end: datetime | None = None


class PortalSessionResponse(BaseModel):
    """Stripe customer portal session response."""

    portal_url: str


class CancelSubscriptionRequest(BaseModel):
    """Request to cancel subscription."""

    cancel_immediately: bool = Field(False, description="Cancel immediately vs at period end")
    reason: str | None = Field(None, description="Cancellation reason")


class CancelSubscriptionResponse(BaseModel):
    """Response after cancellation."""

    success: bool
    message: str
    cancel_at_period_end: bool
    current_period_end: datetime | None = None


# ===========================================
# Payment History
# ===========================================


class PaymentHistoryItem(BaseModel):
    """Single payment history item."""

    id: str
    amount: int  # In cents
    currency: str
    status: str
    description: str | None = None
    invoice_pdf_url: str | None = None
    receipt_url: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class PaymentHistoryResponse(BaseModel):
    """Payment history list response."""

    payments: list[PaymentHistoryItem]
    total_count: int
    has_more: bool


# ===========================================
# Webhooks
# ===========================================


class WebhookEventResponse(BaseModel):
    """Response to webhook events."""

    received: bool = True
    event_type: str | None = None
    message: str | None = None
